[dwsc37] iP

dwsc37 opened the pull request:
# BotManager

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

BotManager frees your mind of having to remember things you need to do. It's,

* text-based
* easy to learn
* ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/dwsc37/ip/releases).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

 - [x] Managing todos
 - [x]  Managing deadlines 
 - [x]  Managing events

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public static void main(String[] args) {
    new BotManager().run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

dwsc37 committed:
feat: Add chatbot with greeting and goodbye

dwsc37 committed:
feat: Add echoing of user input

dwsc37 committed:
feat: Implement adding and listing of tasks

dwsc37 committed:
feat: Add marking and unmarking of tasks

dwsc37 committed:
refactor: Update method names for Task

dwsc37 committed:
feat: Add support for 3 types of tasks

dwsc37 committed:
feat: Add text ui testing

dwsc37 committed:
feat: Add error handling with exceptions

dwsc37 committed:
feat: Add infrastructure for help command

dwsc37 committed:
feat: Add deletion of tasks

dwsc37 committed:
refactor: Refactor command types into enums

dwsc37 committed:
feat: Add full help message

dwsc37 committed:
chore: Update greet message and enum description

dwsc37 committed:
feat: Add infrastructure for saving and loading

dwsc37 committed:
feat: Add saving and loading data from file

dwsc37 committed:
feat: Add handling of missing file and directory

dwsc37 committed:
Merge branch 'branch-Level-7'

dwsc37 committed:
style: Fix indentations

dwsc37 committed:
refactor: Update all dates to use LocalDate type

dwsc37 committed:
chore: Add data folder to gitignore

dwsc37 committed:
Merge branch 'branch-Level-8'

dwsc37 committed:
chore: Clean up code and packages

dwsc37 committed:
Merge branch 'add-gradle-support'

dwsc37 committed:
build: Add basic gradle tasks

dwsc37 committed:
build: Temporarily comment out checkstyle plugin

dwsc37 committed:
build: Add basic JUnit tests

dwsc37 committed:
build: Update archiveBaseName for shadowJar

dwsc37 committed:
chore: Add repo wide javadocs

dwsc37 committed:
Merge branch 'branch-A-JavaDoc'

dwsc37 committed:
build: Add checkstyle to build.gradle

dwsc37 committed:
style: Fix coding style repo wide

dwsc37 committed:
Merge branch 'branch-A-CodingStandard'

dwsc37 committed:
feat: Add searching of tasks by description

dwsc37 committed:
Merge branch 'branch-Level-9'

Incogdino commented on src/main/java/botmanager/Parser.java:
```diff
@@ -0,0 +1,122 @@
+    public Action parseInput(String input) throws BotException {
+        Command command = Command.getCommandFromInput(input.split(" ")[0]);
+        return switch (command) {
+        case LIST -> new ListTasksAction();
```
Great job on this! Many users make the mistake of indenting the case statements.

Incogdino commented on src/main/java/botmanager/BotManager.java:
```diff
@@ -0,0 +1,73 @@
+        while (true) {
+            // read user input
+            String input = ui.readUserInput();
+            if (input.equals(Command.EXIT.getInput())) {
```
The nested function calls might be a bit confusing to read and understand. Perhaps extracting out `Command.EXIT.getInput()` might improve readability?

Incogdino commented on src/main/java/botmanager/Storage.java:
```diff
@@ -0,0 +1,80 @@
+
+    private Task fromData(String data) {
+        String[] args = data.split(" \\| ");
+        return switch (args[0]) {
```
Great job on this! Many users make the mistake of indenting the case statements.

Incogdino commented on src/main/java/enums/Command.java:
```diff
@@ -0,0 +1,72 @@
+        throw new InvalidCommandException(input);
+    }
+
+    private int getMaxUsageLength() {
```
Adding a JavaDoc comment here can help to provide details of the function.

Incogdino commented on src/main/java/enums/Command.java:
```diff
@@ -0,0 +1,72 @@
+/**
+ * Enum representing the different commands available.
+ */
+public enum Command {
```
Great use of the enum class! It helps to extract a lot of details out!

Incogdino reviewed (commented):
LGTM 👍 
Code is inline with the java coding standard defined.

dwsc37 committed:
build: Update build to include javaFx dependencies

dwsc37 committed:
feat: Add GUI for chatbot

dwsc37 committed:
feat: Fully implement GUI for chatboat

dwsc37 committed:
Merge branch 'branch-Level-10'

dwsc37 committed:
chore: Remove text ui testing

dwsc37 committed:
refactor: Refactor method to use vararg

ignatiusoey09 commented on src/main/java/action/Action.java:
```diff
@@ -0,0 +1,30 @@
+import task.TaskList;
+
+/**
+ * Abstract class representing a generic action to be executed by BotManager.
```
Nice JavaDocs!

ignatiusoey09 commented on src/main/java/controller/MainWindow.java:
```diff
@@ -0,0 +1,87 @@
+        String [] messages = Arrays.stream(responses).map(Response::getMessage).toArray(String[]::new);
+        addBotDialog(messages);
+
+        boolean shouldTerminate = Arrays.stream(responses).anyMatch(Response::isTerminating);
```
Great naming!

ignatiusoey09 commented on src/main/java/enums/Command.java:
```diff
@@ -0,0 +1,72 @@
+    private final String usage;
+    private final String description;
+
+    Command(String input, String usage, String description) {
```
Good use of enum attributes

ignatiusoey09 commented on src/main/java/model/Parser.java:
```diff
@@ -0,0 +1,124 @@
+    public Action parseInput(String input) throws BotException {
+        Command command = Command.getCommandFromInput(input.split(" ")[0]);
+        return switch (command) {
+        case LIST -> new ListTasksAction();
```
Didn't know you could write switch cases like that 😮

ignatiusoey09 commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,78 @@
+    /**
+     * Mark the task as done.
+     */
+    protected void mark() {
```
Good use of protected keyword

ignatiusoey09 reviewed (commented):
LGTM! :)

dwsc37 committed:
fix: Update exit function

dwsc37 committed:
refactor: Add assertions to improve code robustness

Task index that are passed into the task list should be in a valid
range. (i.e. from 0 to number of tasks - 1)

This is already checked for in the various action classes. However,
using assertions to ensure this would make the code more robust.

Let's add assertions for the mark, unmark and delete methods in
the TaskList class to do so.

dwsc37 committed:
build: Enable assertions in build.gradle

Java disables assertions by default. However, we want to utilise
assertions in our code during runtime.

Let's update the gradle build config to enable assertions during runtime.

dwsc37 committed:
Merge pull request #1 from dwsc37/branch-A-Assertions

refactor: Add assertions to improve code robustness

dwsc37 committed:
refactor: Removed unused function

One function in BotManager.java is currently unused.

Let's remove it to conform to the code quality standards.

dwsc37 committed:
refactor: Abstract out several string constants

There are several fixed messages that BotManager has (for example welcome and goodbye messages). They currently appear as magic strings in the code.

While the current implementation is fine for now, it may cause problems if these strings need to be reused in other parts of the codebase in the future.

Let's abstract these strings out into as enum class as constants.

dwsc37 committed:
refactor: Remove unused variable

dwsc37 committed:
Merge branch 'master' into branch-A-CodeQuality

dwsc37 committed:
Merge pull request #2 from dwsc37/branch-A-CodeQuality

refactor: Improve code quality

dwsc37 committed:
refactor: Improve stream usage

Currently, we are using if-else statements to check if a Response generated is null.

Let's improve the existing usage of streams to instead filter the Response array for non-null Responses. Let's also refactor the stream usages to more closely follow SLAP.

dwsc37 committed:
Merge branch 'master' into branch-A-Streams

dwsc37 committed:
Merge pull request #3 from dwsc37/branch-A-Streams

refactor: Improve stream usage

dwsc37 committed:
feat: Add command to list schedule

dwsc37 committed:
Merge pull request #4 from dwsc37/branch-BCD-Extension

feat: Add command to list schedule

dwsc37 committed:
build: Setup Github CI

dwsc37 committed:
build: Fix directory structure

dwsc37 committed:
style: Fix formatting of JavaDocs

dwsc37 committed:
feat: Improve GUI and personality

dwsc37 committed:
feat: Improve colour scheme using AI suggestions

dwsc37 committed:
docs: Add showcase image

dwsc37 committed:
fix: Fix schedule command date formatting

dwsc37 committed:
feat: Add handling of invalid file data

dwsc37 committed:
docs: Add user guide

dwsc37 committed:
chore: Fix minor formatting issues

dwsc37 committed:
docs: Update user guide

dwsc37 committed:
Update features overview of user guide

dwsc37 committed:
test: Add unit testing of TaskList class

dwsc37 committed:
style: Remove unused import in ParserTest
